### PR TITLE
salt: Add a  `metalk8s:debug` pillar option

### PR DESCRIPTION
**Pull request by gdemonet**
**Component**: salt, kubernetes

**Context**:

When trying to troubleshoot strange issues, we are often missing detailed logs. Enabling such logs requires manual operations to be performed **after** the initial installation (hence, it's not possible to have said logs during the initial installation).

**Summary**:

With this PR, we introduce a (really crude) solution to the above problem, with a single `debug` toggle in BootstrapConfiguration. When enabled, all the following components will generate more logs:

- Salt master and minions
- containerd
- kubelet
- kube-apiserver, kube-controller-manager, kube-scheduler, kube-proxy
- CoreDNS

_**Note**: for a `single-node-install-rhel` stage, the logs collected on bootstrap go from around 10MB (uncompressed) without the `metalk8s:debug` pillar enabled, to 45MB when it's enabled_

**Acceptance criteria**: 

Keeps working :) there will probably be a performance impact of running everything like this, so for now, it's not documented, and shouldn't be used in production. We'll see where to go from there.

### CODE REVIEW

File: salt/metalk8s/salt/master/files/salt-master-manifest.yaml.j2 [hunk 1]
```diff
@@ -82,7 +81,8 @@ spec:
     - name: salt-api
       image: {{ image }}
       command: ['salt-api']
-      args: ['--log-level=info']
+      args:
+        - --log-level={{ 'debug' if debug else 'info' }}
```
Comment (TeddyAndrieux) [inline salt/metalk8s/salt/master/files/salt-master-manifest.yaml.j2]: Salt-api do not use the value from master config ?

Comment (gdemonet) [reply salt/metalk8s/salt/master/files/salt-master-manifest.yaml.j2]: It's a good question! Not clearly stated in documentation, and I didn't test...

Comment (gdemonet) [reply salt/metalk8s/salt/master/files/salt-master-manifest.yaml.j2]: It does :) Thanks for the tip!

File: salt/metalk8s/container-engine/containerd/files/50-metalk8s.conf.j2 [hunk 2]
```diff
@@ -1,4 +1,7 @@
 [Service]
+# Allow passing arguments to containerd daemon
+ExecStart=
+ExecStart=/usr/bin/containerd {{ args | join(' ') }}
```
Comment (alexandre-allard) [inline salt/metalk8s/container-engine/containerd/files/50-metalk8s.conf.j2]: Not really resilient if there is space in one of the args... but I assume it'll not be the case, at least it's not for now.

File: salt/metalk8s/container-engine/containerd/installed.sls [hunk 3]
```diff
@@ -48,20 +48,23 @@ Create containerd service drop-in:
     - makedirs: true
     - dir_mode: 0755
     - context:
-      environment:
-      {%- if proxies %}
-        {%- set no_proxy = ["localhost", "127.0.0.1"] + networks.values() %}
-        {%- if proxies.no_proxy | default %}
-          {%- do no_proxy.extend(proxies.no_proxy) %}
+        args:
```
Comment (alexandre-allard) [inline salt/metalk8s/container-engine/containerd/installed.sls]: Minor: I would have named this variable `containerd_options` or `containerd_args`,  as `args` is not crystal clear.

File: salt/metalk8s/container-engine/containerd/installed.sls [hunk 4]
```diff
@@ -84,4 +87,7 @@ Configure registry IP in containerd conf:
     - makedirs: true
     - contents: |
         [plugins.cri.registry.mirrors."{{ repo.registry_endpoint }}"]
-          endpoint = ["http://{{ registry_ip }}:{{ registry_port }}"]
+        endpoint = ["http://{{ registry_ip }}:{{ registry_port }}"]
+
+        [debug]
+        level = "{{ 'debug' if metalk8s.debug else 'info' }}"
```
Comment (alexandre-allard) [inline salt/metalk8s/container-engine/containerd/installed.sls:93]: What's the purpose of using both the containerd option `--log-level` and this ? I think we should stick to only one way of doing this, and putting the option in this file seems the better option to me.

File: salt/metalk8s/salt/master/configured.sls [hunk 5]
```diff
@@ -12,6 +13,7 @@ Configure salt master:
     - backup: false
     - template: jinja
     - defaults:
+        debug: {{ metalk8s.debug | tojson }}
```
Comment (alexandre-allard) [inline salt/metalk8s/salt/master/configured.sls]: is the `| tojson` filter really necessary ?

File: eve/main.yml [hunk 6]
```diff
@@ -243,6 +246,7 @@ models:
       name: Create bootstrap configuration file
       env:
         PRODUCT_VERSION: "%(prop:metalk8s_version)s"
+        DEBUG: "%(prop:metalk8s_debug:-false)s"
```
Comment (alexandre-allard) [inline eve/main.yml:251]: Could be cool if all `env` could inherit from a `base_env` which sets this `DEBUG` variable, but it means changing much more stuff, so... nevermind. :)

File: eve/workers/openstack-multiple-nodes/terraform/nodes.tf [hunk 7]
```diff
@@ -92,7 +92,10 @@ resource "openstack_compute_instance_v2" "bootstrap" {
 
   # Generate BootstrapConfiguration
   provisioner "remote-exec" {
-    inline = ["sudo bash scripts/bootstrap-config.sh"]
+    inline = [
+      "sudo chmod +x scripts/bootstrap-config.sh",
```
Comment (alexandre-allard) [inline eve/workers/openstack-multiple-nodes/terraform/nodes.tf:96]: Can't we simply make this script executable directly on the repository, so we don't need to `chmod` it each time.
I don't know if the file provisioner keep the mode of a file, but I think it is.

File: eve/workers/openstack-single-node-rhel/terraform/bootstrap.tf [hunk 8]
```diff
@@ -41,7 +41,8 @@ resource "openstack_compute_instance_v2" "bootstrap" {
 
   provisioner "remote-exec" {
     inline = [
-      "sudo bash scripts/bootstrap-config.sh",
+      "sudo chmod +x scripts/bootstrap-config.sh",
```
Comment (alexandre-allard) [inline eve/workers/openstack-single-node-rhel/terraform/bootstrap.tf:44]: ditto

**Review by alexandre-allard — COMMENTED** [review]
Details

File: salt/metalk8s/container-engine/containerd/files/50-metalk8s.conf.j2 (hunk 2 above)
Comment (gdemonet) [reply salt/metalk8s/container-engine/containerd/files/50-metalk8s.conf.j2]: I can trick it like this:

```jinja
/usr/bin/containerd "{{ args | join('" "'}}"
```

File: salt/metalk8s/container-engine/containerd/installed.sls (hunk 4 above)
Comment (gdemonet) [reply salt/metalk8s/container-engine/containerd/installed.sls:93]: So:
- `--log-level` seems to adjust the main daemon's logging, which ends up in journald
- the `[debug]` config section configures the debug socket, at `/run/containerd/debug.sock` by default (I didn't try listening on this socket so I don't know what gets sent in there :sweat_smile:)

File: salt/metalk8s/salt/master/configured.sls (hunk 5 above)
Comment (gdemonet) [reply salt/metalk8s/salt/master/configured.sls]: I think I had issues with this value being interpreted as a string, but I'll see if I can remove it :)

File: eve/main.yml (hunk 6 above)
Comment (gdemonet) [reply eve/main.yml:251]: Yeaaah... Could be cool to have a better test manager than YAML.

File: salt/metalk8s/container-engine/containerd/files/50-metalk8s.conf.j2 (hunk 2 above)
Comment (gdemonet) [reply salt/metalk8s/container-engine/containerd/files/50-metalk8s.conf.j2]: Done

File: salt/metalk8s/container-engine/containerd/installed.sls (hunk 3 above)
Comment (gdemonet) [reply salt/metalk8s/container-engine/containerd/installed.sls]: Done

File: eve/workers/openstack-multiple-nodes/terraform/nodes.tf (hunk 7 above)
Comment (gdemonet) [reply eve/workers/openstack-multiple-nodes/terraform/nodes.tf:96]: Done

File: eve/workers/openstack-single-node-rhel/terraform/bootstrap.tf (hunk 8 above)
Comment (gdemonet) [reply eve/workers/openstack-single-node-rhel/terraform/bootstrap.tf:44]: Done

File: salt/metalk8s/container-engine/containerd/installed.sls (hunk 4 above)
Comment (alexandre-allard) [reply salt/metalk8s/container-engine/containerd/installed.sls:93]: Oh ok, I see the difference, I thought it was the same thing, I've checked the man (containerd-config.toml) and effectively it is not related.
Still, that's weird that there is no option to change daemon log level within this configuration file.

File: salt/metalk8s/container-engine/containerd/files/50-metalk8s.conf.j2 [hunk 9]
```diff
@@ -1,4 +1,11 @@
 [Service]
+# Allow passing arguments to containerd daemon
+ExecStart=
+{%- if containerd_args %}
+ExecStart=/usr/bin/containerd "{{ args | join('" "') }}"
```
Comment (alexandre-allard) [inline salt/metalk8s/container-engine/containerd/files/50-metalk8s.conf.j2]: Don't forget to also change the name here:
```suggestion
ExecStart=/usr/bin/containerd "{{ containerd_args | join('" "') }}"
```

Comment (gdemonet) [reply salt/metalk8s/container-engine/containerd/files/50-metalk8s.conf.j2]: :facepalm:

File: salt/metalk8s/salt/master/configured.sls (hunk 5 above)
Comment (gdemonet) [reply salt/metalk8s/salt/master/configured.sls]: Removed!

File: eve/main.yml [hunk 10]
```diff
@@ -1103,7 +1109,9 @@ stages:
             ISO_PATH: /tmp/metalk8s.iso
       - ShellCommand:
           <<: *bootstrap_config
-          env: *_env_version_prev
+          env:
+            <<: *_env_version_prev
+            DEBUG: "%(prop:metalk8s_debug)s"
```
Comment (alexandre-allard) [inline eve/main.yml]: No default value to fallback on, as you did above:
```suggestion
            DEBUG: "%(prop:metalk8s_debug:-false)s"
```

File: eve/main.yml [hunk 11]
```diff
@@ -1408,7 +1416,9 @@ stages:
             ISO_PATH: /tmp/metalk8s.iso
       - ShellCommand:
           <<: *bootstrap_config
-          env: *_env_promoted_version
+          env:
+            <<: *_env_promoted_version
+            DEBUG: "%(prop:metalk8s_debug)s"
```
Comment (alexandre-allard) [inline eve/main.yml]: ditto
```suggestion
            DEBUG: "%(prop:metalk8s_debug:-false)s"
```

File: eve/main.yml (hunk 10 above)
Comment (gdemonet) [reply eve/main.yml]: Done

File: eve/main.yml (hunk 11 above)
Comment (gdemonet) [reply eve/main.yml]: Done

**Review by alexandre-allard — APPROVED** [review]
LGTM, brace yourselves, debugging logs are coming.

**Pull request merged**